### PR TITLE
Fix for #1989 Replace modulo with fastmodulo in HashTable

### PR DESCRIPTION
**Pull request by ErhanAtesoglu**
This pull request contains two commits.  First commit Replaces modulo with FastMod.  The second commit updates thrown exceptions to use ThrowHelper methods.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=55467) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=55467) before we can accept your contribution.<br/><br/>:x: ErhanAtesoglu [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=55467)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=55467) it.</sub>

File: src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs [hunk 1]
```diff
@@ -730,6 +757,9 @@ private void rehash(int newsize)
             //   2) Protect against an OutOfMemoryException while allocating this
             //      new bucket[].
             bucket[] newBuckets = new bucket[newsize];
+#if TARGET_64BIT
+            _fastModMultiplier = HashHelpers.GetFastModMultiplier((uint)newsize);
+#endif
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs:762]: This is not safe.  One of the key reasons code still uses Hashtable is its threading guarantees, which enable any number of reads to execute concurrently with one writer.  By updating the _fastModMultiplifer here, concurrent readers could start to fail.

File: src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs [hunk 2]
```diff
@@ -258,7 +262,7 @@ public Hashtable(int capacity) : this(capacity, 1.0f)
         public Hashtable(int capacity, float loadFactor)
         {
             if (capacity < 0)
-                throw new ArgumentOutOfRangeException(nameof(capacity), SR.ArgumentOutOfRange_NeedNonNegNum);
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.capacity, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs:265]: What's wrong with keeping `nameof` as the argument?

File: src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs [hunk 3]
```diff
@@ -267,13 +271,16 @@ public Hashtable(int capacity, float loadFactor)
 
             double rawsize = capacity / _loadFactor;
             if (rawsize > int.MaxValue)
-                throw new ArgumentException(SR.Arg_HTCapacityOverflow, nameof(capacity));
+                ThrowHelper.ThrowArgumentException(ExceptionResource.Arg_HTCapacityOverflow, ExceptionArgument.capacity);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs:274]: Nit: I would prefer it if we kept these enhancements in a separate PR from the fastmod changes.

**Comment by eiriktsarpalis** [comment]
Do we generally accept these types of enhancements for non-generic collections? I would assume we keep them around primarily for back-compat reasons so any change should require substantial justification.

**Comment by stephentoub** [comment]
> I would assume we keep them around primarily for back-compat reasons so any change should require substantial justification.

In general, that's true.  `Hashtable` is a _little_ special, in that it provides some thread-safety guarantees that `Dictionary<>` doesn't, namely you can have any number of readers concurrent with a single writer, and so we still use it in a few places where that guarantee is helpful (typically this is just corelib, as higher-level code could use ConcurrentDictionary that also supports that and more).

That said, Hashtable is already much more expensive than Dictionary, and it's not clear to me that a) this will meaningfully move the needle on its performance, and b) that even if it does on microbenchmarks, it would meaningfully move the needle on the higher-level APIs that still need to consume it and for which perf could matter.

@ErhanAtesoglu, thanks for your contribution here.  Can you share an example of a type in the core libraries that still uses Hashtable and where this change measurably helps improve perf of that type?

**Comment by danmoseley** [comment]
It would also be good if you could share microbenchmark results for HashTable too. Please see the dotnet/performance repo which has tests and excellent instructions for getting numbers with and without a change like this.

**Comment by danmoseley** [comment]
With respect to whether to table changes to the legacy collections, another consideration is how much existing code uses them. We can easily gather data on that from Nuget if we wish. I'm going to guess it's considerable in which case I think we'd take a change if it have significant perf at minimal risk/churn.

**Comment by eiriktsarpalis** [comment]
FWIW here are the instructions on how to benchmark your own branch: https://github.com/dotnet/performance/blob/main/src/benchmarks/micro/README.md#private-runtime-builds

**Comment by ErhanAtesoglu** [comment]
Just my quick thoughts.  After testing it through the gauntlet I noticed I made some errors along the way.  First I was using fastmod on longs which is not possible.  Second with the uints I'm still getting test errors on what should be safe uses of FastMod.  I have two acceptable lines where the replacement doesn't break any of the tests, but since similar lines do break the tests I'm not confident of these changes.

My gut feeling says,  leave HashTable alone.

**Comment by stephentoub** [comment]
> My gut feeling says, leave HashTable alone.

Ok.  Thanks for your efforts here.  I'm going to close this.

File: src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs (hunk 2 above)
Comment (nelsondani) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Hashtable.cs:265]: 2566346[[_**234567**_]()]()

**Pull request closed without merging**